### PR TITLE
Reduce input during on-boarding

### PR DESCRIPTION
**Pull request by TheSin-**
- Add mdns and ngx-electron plugin
    - Add electron-rebuild to dev depends so that electron can be rebuilt with the installed node
    - Fix setTimeout and setInstance to with with Node modules (required for mdns)
    - Reduce the version on node as electron needs to be on the same version for this to work
    - Change order in on boarding wizard to select API first
    - Use mdns to detect and list octoprint instances (WIP need to figure out min version required)
    - Auto set Printer name if a detected install if found.

Step 1 in #921

### CODE REVIEW

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [inline src/app/config/no-config/no-config.component.ts:35]: I guess we should use octoprintInstance and octoprintApiMessage here, since the full name is used in other vars as well.

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [inline src/app/config/no-config/no-config.component.ts:16]: the `:number` is redundant here

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [inline src/app/config/no-config/no-config.component.ts:48]: We should stick with port 5000 here, since that is the official port from OctoPrint. (80 is just added from OctoPi)

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [inline src/app/config/no-config/no-config.component.ts:173]: Can we put this in `main.js`, this would remove the ngx-electron dependency and the icp links are already there, plus we would get around having to add node to the main angular app.

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [inline src/app/config/no-config/no-config.component.ts:203]: Should return `Promise<void>`
```suggestion
  public async validateConfig(): Promise<void> {
```

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [inline src/app/config/no-config/no-config.component.ts:221]:
```suggestion
  public async increasePage(): Promise<void> {
```

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [inline src/app/config/no-config/no-config.component.ts:176]:
```suggestion
  public async testOctoprintAPI(): Promise<boolean> {
```

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [inline src/app/config/no-config/no-config.component.ts:255]: 👍

File: src/app/config/no-config/no-config.component.html
Comment (UnchartedBull) [inline src/app/config/no-config/no-config.component.html:44]: I think these should be sorted again once everything is complete. Fine for now though, since the whole setup screen needs a revamp once all the changes are in.

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [inline src/app/config/no-config/no-config.component.ts:27]: I think we can be more inclusive here. I would propose >1.3.5.

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [inline src/app/config/no-config/no-config.component.ts:247]: This will query the API 4 times (for each page after 2). I would prefer to just have the check for connection at the very end, where it is currently located. Or are there any benefits to having it here?

**Review by UnchartedBull — CHANGES_REQUESTED** [review]
Looks really good. The only bigger point to discuss here, is whether the mdns part should be moved to the main.js, where we already have full access to the node api, sharing data can be easily done via the electron IPC. And the API Connection check can be at the end I guess. Everything else is just some minor codestyle stuff (I guess the pipeline will also complain about some of that).

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [reply src/app/config/no-config/no-config.component.ts:173]: Let me know if you want me to do this.

File: src/app/config/no-config/no-config.component.ts
Comment (TheSin-) [reply src/app/config/no-config/no-config.component.ts:35]: Sure, I'm lazy I guess ;). I'll update this tonight.

File: src/app/config/no-config/no-config.component.ts
Comment (TheSin-) [reply src/app/config/no-config/no-config.component.ts:16]: I like it for consistency, but I can't remove it if it bugs ya.

File: src/app/config/no-config/no-config.component.ts
Comment (TheSin-) [reply src/app/config/no-config/no-config.component.ts:48]: oh I didn't know that, maybe we could put a note, but since you official release used is octopi it's make more sense if it worked out of the box I know I got stuck on this a lot in the start.

File: src/app/config/no-config/no-config.component.ts
Comment (TheSin-) [reply src/app/config/no-config/no-config.component.ts:173]: Yeah that is beyond me, this took me long enough to figure out ;)

File: src/app/config/no-config/no-config.component.ts
Comment (TheSin-) [reply src/app/config/no-config/no-config.component.ts:203]: Didn't know I could do this void there, I know there are other types but I didn't know void was one, this will be handy thanks.

File: src/app/config/no-config/no-config.component.ts
Comment (TheSin-) [reply src/app/config/no-config/no-config.component.ts:255]: Did I miss this one?

File: src/app/config/no-config/no-config.component.html
Comment (TheSin-) [reply src/app/config/no-config/no-config.component.html:44]: I need to go retest these I though I tested going all the way through and back 👍 

But that is why I didn't sort it I assume most of this will be gone in step 2.

File: src/app/config/no-config/no-config.component.ts
Comment (TheSin-) [reply src/app/config/no-config/no-config.component.ts:27]: sure I had no idea as I haven't finished going through everything, since we plan to use push when was that added?

File: src/app/config/no-config/no-config.component.ts
Comment (TheSin-) [reply src/app/config/no-config/no-config.component.ts:247]: it won't cause of line 234.  Once it connects once it'll stop.

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [reply src/app/config/no-config/no-config.component.ts:16]: Screenshot 2020-08-31 at 15 21 43

Actually bugs VSCode :D

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [reply src/app/config/no-config/no-config.component.ts:255]: No that was just an: I like that reaction, sorry for the confusion :)

File: src/app/config/no-config/no-config.component.html
Comment (UnchartedBull) [reply src/app/config/no-config/no-config.component.html:44]: Leave that for now, We can go through this once the rest is done. There will be a couple of changes here I guess.

File: src/app/config/no-config/no-config.component.ts
Comment (TheSin-) [reply src/app/config/no-config/no-config.component.ts:16]: ahh I use VIM so I don't have that issue, seems like a VSCode issue, but we can remove it since it doesn't hurt anything but my eyes ;)

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [reply src/app/config/no-config/no-config.component.ts:247]: Oh I missed that if statement. Sorry.

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [reply src/app/config/no-config/no-config.component.ts:173]: Ok I'll take that one :)

File: src/app/config/no-config/no-config.component.ts
Comment (UnchartedBull) [reply src/app/config/no-config/no-config.component.ts:48]: A note would be a good idea yeah. Although localhost:5000 should also work on OctoPi. OctoPi is just creating a reverse proxy from :80 to :5000 AFAIK.

Comment (TheSin-) [reply src/app/config/no-config/no-config.component.ts:48]: Odd it doesn't work on my Octodash install or my 2 other octopi installs, maybe I'm doing something wrong, I'll look into it.

Either way I'll add a note for it.

File: src/app/config/no-config/no-config.component.ts
Comment (TheSin-) [reply src/app/config/no-config/no-config.component.ts:27]: looks like GitHub botched up my comments.  This comment was for an other one.

For this one I mentioned something about when push was added to the api since we will be using that.  I still need to add a version compare lib like semver or something.  I can do that tonight too.

**Comment by TheSin-** [comment]
Because of my changes to post install I might need to move the electron-rebuild dev depend into depends?  I'm really new to electron.

**Comment by TheSin-** [comment]
Done addressing things, I'm going to see if I can figure out why CI is falling next.

**Comment by UnchartedBull** [comment]
Sorry for the spam commits, just wanted to test whether I can push to the PR :)

**Comment by TheSin-** [comment]
no problem I've been learning what can be done too ;)

**Comment by TheSin-** [comment]
The new Travis fails is due to the changes in .eslintrc.js

Which is out side of my knowledge

https://travis-ci.com/github/UnchartedBull/OctoDash/jobs/379614533

**Comment by UnchartedBull** [comment]
Yeah I'll fix those :)

**Comment by TheSin-** [comment]
I think we committed at the same time and some how mine because forced even though I didn't force it :\

**Comment by UnchartedBull** [comment]
No problem, you only did linter stuff right? I'll probably just force push over that, if that's ok :)

**Comment by TheSin-** [comment]
Sorry was driving home, nice changes I'm going to test it after dinner.

**Comment by TheSin-** [comment]
it doesn't test the connect, thus I can't pre populate the rest of the config anymore nor will it warn.  The changes look much better but they user experience isn't as good.

**Comment by TheSin-** [comment]
It's also very slow for initial detection now, and looking at it I don't think it's continue to search so if you turn it one once you realize you need it it'll be too late.

**Comment by UnchartedBull** [comment]
It currently doesn't test the connect no. The idea I'm having currently is to now trigger the API Connection with the popup thingy. This will be easiest if we use the OctoPrint JS Client, which would also allow for super easy socket connection down the line. It's just a big pain to include a dynamically loaded JS library into a TS Angular application. Once that it working we would have the OctoPrint instance + an authorised client without having to put any text input anywhere (of course there will be fallbacks).

The idea for now is:
- Page 1: List nodes / enter node manually
- Page 2: Authorize via popup in WebUI / enter token manually
- Page 3 and onwards: a fully authenticated OctoPrint Client will be ready and can be used to pre-populate most of the values.

I don't know if that makes sense, but I hope so. It's 1AM here and I've been coding since 9AM (including work though), so my brain is a bit mushy right now :D But would that work for you? I know I shifted around a lot of code, but this also means that we have node.js back out of the main Angular app.

Initial detection is just slow due to Angulars Change Detection, we could trigger those manually and have the data almost instantly. I might add that tomorrow. It will continue to search. Each time you leave page 1 the browser will be stopped to not spam the whole network with mdns requests, but if page 1 get's opened again it will search again. I tried that with turning the WiFi off, which makes the instance disappear, but it comes back once the connection is reestablished.

**Comment by UnchartedBull** [comment]
Ok I updated the change detection, should be as fast as before now. Sorry again if I crushed your plans here, but I think we do it now, we should do it right and since we need that library later anyways we might include it now. If you want some changes let me know and if you have more questions let me know as well :) Otherwise enjoy your dinner!

**Comment by TheSin-** [comment]
ahh that was going to be more phase 2 ;). You are doing it in one sweep, well I'm excited to see it, your UI design is beautiful I love that part a lot.

**Comment by TheSin-** [comment]
That's totally okay, I just didn't know you were going to crush it all out at once.  I didn't want to make the PR too large.  I've done that in other projects and it's been over 10 years and it's still not accepted.  So I'm shy about it now, but that is the end goal so I'm more then happy to help get us there any way I can cause I think that will be the ultimate user experience!

**Comment by UnchartedBull** [comment]
I didn't plan to do it from the beginning but then I kinda got lost in it. I'll probably can do that tomorrow evening/night :)

I'm more than ok with large PRs for OctoDash (I do them myself quite frequently), so don't be shy there :) it might just take a little longer for me to review them.

And yes I'm already loving that discovery so much. Makes it a lot easier for not so tech savvy users and is also a great learning for me.

**Comment by TheSin-** [comment]
Well I'm excited to say the least, I'm going to try to tackle the push conversion next, but I'll wait for this PR to be done so I don't need to make a branch ;). Let me know if you need anything else form me on this other then testing.

**Comment by TheSin-** [comment]
looking at the changes we might not need any of the electron-rebuild stuff, but I haven't tested that yet, I'll merge in master if you'd like so that there aren't any conflicts.

**Comment by UnchartedBull** [comment]
Sure! I hope to get this done by tomorrow, after that you should have the authorized client available and then you can pre-fill the values based on printer profile / other information. Rebuild probably isn't needed anymore. I can remove that :)

**Comment by TheSin-** [comment]
I'll test it tonight and remove it that way you can focus on more useful parts that I can't do ;)

**Comment by TheSin-** [comment]
No problems removing it all, just going to see if the travis changes were needed for mdns in general or just the rebuild.

**Comment by TheSin-** [comment]
Just to keep track in my testing one other piece the tis missing from my code is that you do not reset values on back.

so if I pick an Instance from the list, then go back cause I changed my mind and hit enter manually, it's the info from the last Instance I tried instead of the default local value which it should be IMHO.

**Comment by UnchartedBull** [comment]
Ok quick update: the reset if manually entered you're talking about has been implemented. Checking that OctoPrint is running on the specified host:port is also done and the library is loaded correctly. So there is just the authentication workflow left. I hoped that I could do that today, the freakin' jQuery issue took me a good hour to solve though. Gotta postpone the authentication to tomorrow, sorry. After that everything should be done from my side for now I guess. Maybe some adjustment and deletion of non-necessary fields (due to printer profiles.)

My game plan here is to merge this into a separate branch for now and then do one sweep from that branch into master once push updates have been implemented and the setup has been finalised. Will be released with 2.1.0 then I guess, so there'll probably be one or two patch release before this makes it into the final release. Does that work for you?

**Comment by TheSin-** [comment]
Looking good...

There was a small bug that bugged me, on page 1 it defaults to search but if you clicked next it used localhost even though it wasn't listed or selected.  I removed the next button forcing the user to select an Instance or select local to correct the flow and avoid errors.

**Comment by TheSin-** [comment]
There was also a small bug that allowed you to skip past page 0 which I corrected.  Now that I think about it I should have done it in the functions as well to be safe, an other commit inc

**Comment by TheSin-** [comment]
nice that is how I was going to move it but I didn't want to move the click incase there was a reason it was on the div and not the span.

**Comment by UnchartedBull** [comment]
Ok, great news. Seems like everything is working now. Connection is being established and successfully authorized (for now it just retrieves the printer name). No keyboard needed anymore for the setup 🎉 . Let me know if you find any bugs / have some ideas for further changes, this should be ready to merge otherwise :)

Screenshot 2020-09-04 at 18 33 09

**Comment by TheSin-** [comment]
Awesome can't wait to try/test it. I'm just heading out for a bit, I'll test it when I return though!

**Comment by TheSin-** [comment]
I'm searching like made but in electron is there no way to show the number input spinner ?  That would be the best to keep it 100% touch only.

**Comment by TheSin-** [comment]
Maybe something like http://angular-step-input.10kb.nl ?

**Comment by TheSin-** [comment]
I have an idea for the numeric inputs that I"m going to try to work on this week.

**Comment by TheSin-** [comment]
This is my idea, but the two way data bind is really really slow to update and I can't figure out why.  It needs to be much faster to be usable, is this due to electron?

**Comment by TheSin-** [comment]
It's something with the IPC calls, if I set page to 4 and start it (skip the api calls) then it's super responsive and very nice to use.  If I start at page 0 then it's pretty much unusable.  Since I really do not understand the IPC stuff I think you might need to debug it.

**Comment by UnchartedBull** [comment]
Hi, sorry last week has been kinda busy. I'll have a look at that today or tomorrow :) Is there an option to enter text manually though? I think it will be really hard to get to the right number on 3.5" device. I guess we can change that now as well but all other text fields should be updates as well to the new input method. That will be tracked in #762 though.

**Comment by TheSin-** [comment]
I figured we could make the number display (span) into an input, or we could have a toggle to go between the 2.  But I didn't want to work on it if I couldn't get it working at all.

**Comment by UnchartedBull** [comment]
Just tried it and it works just fine for me. No noticable lag at all (I started on page 0). It looks nice and certainly is a super clean way of inputting a number. The main problem here is, that some (possibly many) people are using 3.5" GPIO displays with something close to 1 or 2Hz refresh rate. This makes nice gestures for essential stuff almost impossible.

I redid those so we have + und - buttons (hold them to go up/down faster). I'm fairly happy with that, but please provide feedback on that. Sorry for crashing everything again. This PR is ready to be merged into master from my side now (if the pipeline passes). Let me know what you're thinking :)

**Comment by TheSin-** [comment]
I'll try them in a few minutes, just have a few meetings first.  But if it works well I think that is the last of it, I'm excited!!  This is a great update indeed!

**Comment by UnchartedBull** [comment]
Sure take your time, doesn't really matter if this PR stays open for another day or two :)

**Comment by TheSin-** [comment]
- Looks good, is the alignment issue on purpose or is it just cause I'm running it on my Mac?

- Are we going to add it to the Interval on Page 5 as well?  Just to keep things consistent?

Screen Shot 2020-09-15 at 8 51 28 AM

**Comment by UnchartedBull** [comment]
That layout issue shouldn't happen. Some weird CSS thing going on with the input I guess. Should be fixed now though. The value interval won't receive the buttons since it will be removed soon (once the push updates are in place there is really no need for that anymore), so we can safe ourself the hassle of adding the controller there to.

**Comment by TheSin-** [comment]
yeah I was on the fence about cause really it shouldn't normally be changed.  I'll try the new changes but otherwise I'm gtg on this, it's amazing work!

**Comment by TheSin-** [comment]
Odd it fixed the large version but not the smaller version

Screen Shot 2020-09-15 at 9 10 55 AM

Screen Shot 2020-09-15 at 9 11 04 AM

**Comment by UnchartedBull** [comment]
Damn, though I had fixed it. Will try again later today, once that's fixed we're good to merge.

**Comment by TheSin-** [comment]
I believe so!!

**Comment by TheSin-** [comment]
Fixed, for me in both views now.

**Comment by UnchartedBull** [comment]
🚢

**Pull request merged**
